python/lib/core/dmod/core/common/tasks.py

```diff
@@ -0,0 +1,180 @@
+    incomplete_message: TASK_MESSAGE = None,
+    cancel_failed_message: TASK_MESSAGE = None
+):
+    """
+    Attempt to cancel a specfic task
+
+    Args:
+        task: The task to cancel
+        cancel_message: A message or function to create a message to report when cancelling a task
+        incomplete_message: A message or function to generate a message for when a task could not be properly canceled in time
+        cancel_failed_message: A message to report in an exception if an error is encountered when cancelling a task
+    """
+
+    # Only try to cancel the task if it hasn't completed yet
+    if not (task.done() or task.cancelled()):
+        try:
+            # Evaluate the cancel message if it's a function
+            if cancel_message is None:
+                cancel_message_for_task = f"Cancelling {task.get_name()}"
+            elif inspect.iscoroutinefunction(cancel_message):
+                cancel_message_for_task = await cancel_message(task)
+            elif isinstance(cancel_message, typing.Callable):
+                cancel_message_for_task = cancel_message(task)
+            else:
+                cancel_message_for_task = cancel_message
+
+            task.cancel()
+
+            completed = await wait_on_task(task)
+
+            # Python 3.9+ lets you pass the message into cancel, but not 3.8, so we do this manually
+            if task.cancelled():
+                logging.debug(cancel_message_for_task)
+
+            # Report an error and move on if the cancel function did not properly stop execution
+            if not completed:
+                # Evaluate the incomplete message if the given message is a function
+                if incomplete_message is None:
+                    task_incomplete_message = f"Could not cancel {task.get_name()} within the given amount of time"
+                elif inspect.iscoroutinefunction(incomplete_message):
+                    task_incomplete_message = await incomplete_message(task)
+                elif inspect.isroutine(incomplete_message):
+                    task_incomplete_message = incomplete_message(task)
+                else:
+                    task_incomplete_message = incomplete_message
+
+                logging.error(task_incomplete_message)
+        except BaseException as cancel_exception:
+            if cancel_failed_message is None:
+                cancel_failed_message = f"Failed to cancel an incomplete task: " \
+                                        f"{task.get_name()}; {str(cancel_exception)}"
+            elif inspect.iscoroutinefunction(cancel_failed_message):
+                cancel_failed_message = await cancel_failed_message(task, cancel_exception)
+            elif isinstance(cancel_failed_message, typing.Callable):
+                cancel_failed_message = cancel_failed_message(task, cancel_exception)
+
+            logging.error(cancel_failed_message, exc_info=cancel_exception)
+
+
+async def cancel_tasks(
```

Comment:
ditto, you could return a tuple of booleans that indicate the success or failure of cancellation. That might not make as much sense here given that you could pass in an unordered iterable of `tasks`.

Reply:
See above; it's going to return a tuple of results stating what the task name was, if it was successfully ended, and any attached messaging.

Reply:
Right, so like above, `cancel_tasks` just needs to return the `await asyncio.gather`